runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java

```diff
@@ -102,11 +121,370 @@ public void visitPrimitiveTransform(Node node) {
       // TODO(JIRA-5649): Don't even emit these transforms in the generated protos.
       res = elideDeprecatedViews(res);
     }
+
+    ExternalTranslationOptions externalTranslationOptions =
+        pipeline.getOptions().as(ExternalTranslationOptions.class);
+    List<String> urnsToOverride = externalTranslationOptions.getTransformsToOverride();
+    if (urnsToOverride.size() > 0) {
+      // We use PTransformPayloadTranslators and the Transform Service to re-generate the pipeline
+      // proto components for the updated transform and update the pipeline proto.
+      Map<String, PTransform> transforms = res.getComponents().getTransformsMap();
+      List<String> alreadyCheckedURns = new ArrayList<>();
+      for (Entry<String, PTransform> entry : transforms.entrySet()) {
+        String urn = entry.getValue().getSpec().getUrn();
+        if (!alreadyCheckedURns.contains(urn) && urnsToOverride.contains(urn)) {
+          alreadyCheckedURns.add(urn);
```

Comment:
Won't this only override one transform of each type in the graph? (Seems worth a test too.)

Reply:
Oh, I see, for each transform you're then checking for any other transforms with that URN. I think it'd be cleaner to loop over URNs to translate rather than over transforms.entrySet() to get the set of suitable URNs. .